streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,83 +78,123 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
      */
-    public void makeReady(final Map<InternalTopicConfig, Integer> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                final Map<InternalTopicConfig, Integer> topicsToBeCreated = validateTopicPartitions(topics, existingTopicPartitions);
-                if (topicsToBeCreated.size() > 0) {
-                    if (metadata.brokers().size() < replicationFactor) {
-                        throw new StreamsException("Found only " + metadata.brokers().size() + " brokers, " +
-                            " but replication factor is " + replicationFactor + "." +
-                            " Decrease replication factor for internal topics via StreamsConfig parameter \"replication.factor\"" +
-                            " or add more brokers to your cluster.");
+    public void makeReady(final Map<String, InternalTopicConfig> topics) {
+        final Map<String, Integer> existingTopicPartitions = getNumPartitions(topics.keySet(), true);
+        final Set<InternalTopicConfig> topicsToBeCreated = validateTopicPartitions(topics.values(), existingTopicPartitions);
+        if (topicsToBeCreated.size() > 0) {
+            final Set<NewTopic> newTopics = new HashSet<>();
+
+            for (final InternalTopicConfig internalTopicConfig : topicsToBeCreated) {
+                final Properties topicProperties = internalTopicConfig.toProperties(windowChangeLogAdditionalRetention);
+                final Map<String, String> topicConfig = new HashMap<>(defaultTopicConfigs);
+                for (final String key : topicProperties.stringPropertyNames()) {
+                    topicConfig.put(key, topicProperties.getProperty(key));
+                }
+                newTopics.add(
+                    new NewTopic(
+                        internalTopicConfig.name(),
+                        internalTopicConfig.numberOfPartitions(),
+                        replicationFactor)
+                    .configs(topicConfig));
+            }
+
+            final CreateTopicsResult createTopicsResult = adminClient.createTopics(newTopics);
+
+            final Map<String, KafkaFuture<Void>> createTopicResults = createTopicsResult.values();
+            for (final Map.Entry<String, KafkaFuture<Void>> createTopicResult : createTopicResults.entrySet()) {
+                try {
+                    createTopicResult.getValue().get();
+                } catch (final ExecutionException couldNotCreateTopic) {
+                    final String topicName = createTopicResult.getKey();
+                    if (couldNotCreateTopic.getCause() instanceof TopicExistsException) {
+                        verifyIfTopicExistsWithCorrectNumberOfPartitions(topicName, topics.get(topicName));
```

Comment:
Hmm.. do we want handle this case by having a second check on the topics that failed to be created? This is one round-trip for each existed-and-hence-failed to create topic to verify.

The cases that the existed topic has a different num.partitions would be 1) it exists with different number of partitions, and is being deleted but not yet completed; or 2) another leader has been elected and a new rebalance triggered with different num source topic partitions and hence this broker has already been likely kicked.

For both cases it is not worth double checking the num.partitions again, so I feel just treat is as OK is fine?